lib/src/config.cc

```diff
@@ -304,6 +310,130 @@ namespace hocon {
+        boost::algorithm::trim(input);
+        string original_unit_string = boost::algorithm::trim_left_copy_if(input, !boost::algorithm::is_alpha());
+        string unit_string = original_unit_string;
+        string number_string = boost::algorithm::trim_copy(input.substr(0, input.length() - unit_string.length()));
+
+        if (number_string.empty()) {
+            throw bad_value_exception(*origin_for_exception, path_for_exception, "No number in duration value '" + input + "'");
+        }
+
+        if (unit_string.length() > 2 && unit_string.back() != 's') {
+            unit_string += "s";
+        }
+
+        try {
+            int64_t number = boost::lexical_cast<int64_t>(number_string);
+            if (unit_string == "ns" || unit_string == "nanos" || unit_string == "nanoseconds") {
+                return nanoseconds(number);
+            } else if (unit_string == "us" || unit_string == "micros" || unit_string == "microseconds") {
+                return microseconds(number);
+            } else if (unit_string.empty() || unit_string == "ms" || unit_string == "millis" || unit_string == "milliseconds") {
+                return milliseconds(number);
+            } else if (unit_string == "s" || unit_string == "seconds") {
+                return seconds(number);
+            } else if (unit_string == "m" || unit_string == "minutes") {
+                return minutes(number);
+            } else if (unit_string == "h" || unit_string == "hours") {
+                return hours(number);
+            } else if (unit_string == "d" || unit_string == "days") {
+                return hours(number * 24);
+            } else {
+                throw bad_value_exception(*origin_for_exception, path_for_exception,
+                    "Could not parse time unit '" + original_unit_string + "' (try ns, us, ms, s, m, h, or d)");
+            }
+        } catch (boost::bad_lexical_cast& ex) {
+            try {
+                double decimal = boost::lexical_cast<double>(number_string);
+                if (unit_string == "ns" || unit_string == "nanos" || unit_string == "nanoseconds") {
+                    // do nothing, we want to use the number as nanoseconds
+                } else if (unit_string == "us" || unit_string == "micros" || unit_string == "microseconds") {
+                    decimal *= 1000.0;
+                } else if (unit_string.empty() || unit_string == "ms" || unit_string == "millis" || unit_string == "milliseconds") {
+                    decimal *= 1000000.0;
+                } else if (unit_string == "s" || unit_string == "seconds") {
+                    decimal *= 1000000000.0;
+                } else if (unit_string == "m" || unit_string == "minutes") {
+                    decimal *= 60000000000.0;
+                } else if (unit_string == "h" || unit_string == "hours") {
+                    decimal *= 3600000000000.0;
+                } else if (unit_string == "d" || unit_string == "days") {
```

Comment:
This has so many overflow possibilities, but Im not really sure what to do with it given the interface we're provided. If you're using decimal time then it's your problem.